File: send.py
import socket
import os
import hashlib
import struct
import threading
import sys
import time

PAYLOAD_LENGTH = 1024
PACKET_SIZE = PAYLOAD_LENGTH + 16 + 32
WINDOW_SIZE = 8
ACK_SIZE = 6
TIMEOUT = 5


def send_file(file_path, ip, port):
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.settimeout(1)

    transmission_id = int.from_bytes(os.urandom(2), "big")
    file_size = os.path.getsize(file_path)
    file_name = os.path.basename(file_path).encode("utf-8")
    max_sequence_number = (
        file_size // PAYLOAD_LENGTH + ((file_size % PAYLOAD_LENGTH) != 0) + 1
    )

    # Send initial packet with file name and max sequence number
    sequence_number = 0
    zero_packet = (
        struct.pack("!HII", transmission_id, sequence_number, max_sequence_number)
        + file_name
    )
    while True:
        sock.sendto(zero_packet, (ip, port))
        try:
            ack, _ = sock.recvfrom(ACK_SIZE)
            ack_transmission_id, ack_sequence_number = struct.unpack("!HI", ack[:6])
            if (
                ack_transmission_id == transmission_id
                and ack_sequence_number == sequence_number
            ):
                print(f"ack: {ack_sequence_number}")
                break
        except socket.timeout:
            continue
    sequence_number += 1

    # Sliding window variables
    base = 1
    lock = threading.Lock()
    locked = threading.Condition(lock)
    packets = {}
    acked = [False] * max_sequence_number

    def send_packets():
        nonlocal sequence_number
        with open(file_path, "rb") as f:
            f.seek((base - 1) * PAYLOAD_LENGTH)
            while base < max_sequence_number:
                with lock:
                    while (
                        sequence_number < base + WINDOW_SIZE
                        and sequence_number < max_sequence_number
                    ):
                        f.seek((sequence_number - 1) * PAYLOAD_LENGTH)
                        data = f.read(PAYLOAD_LENGTH)
                        packet = (
                            struct.pack("!HI", transmission_id, sequence_number) + data
                        )
                        packets[sequence_number] = packet
                        sock.sendto(packet, (ip, port))
                        sequence_number += 1
                with locked:
                    locked.wait(timeout=3)

    def receive_acks():
        nonlocal base
        ack_sequence_number = 0
        while base < max_sequence_number:
            try:
                start = time.time()
                while base < sequence_number and base < max_sequence_number:
                    now = time.time()
                    if now - start > 1:
                        break
                    ack, _ = sock.recvfrom(ACK_SIZE)
                    ack_transmission_id, ack_sequence_number = struct.unpack(
                        "!HI", ack[:6]
                    )
                    if ack_transmission_id == transmission_id:
                        with lock:
                            if not acked[ack_sequence_number]:
                                acked[ack_sequence_number] = True
                                packets.pop(ack_sequence_number)
                                while base < max_sequence_number and acked[base]:
                                    base += 1
            except socket.timeout:
                for seq in range(base, sequence_number):
                    if not acked[seq]:
                        sock.sendto(packets[seq], (ip, port))
            with locked:
                locked.notify_all()

    send_thread = threading.Thread(target=send_packets)
    ack_thread = threading.Thread(target=receive_acks)
    send_thread.start()
    ack_thread.start()
    send_thread.join()
    ack_thread.join()

    # Send final packet with MD5 checksum
    md5_checksum = calculate_md5(file_path)
    final_packet = struct.pack("!HI", transmission_id, sequence_number) + md5_checksum
    while True:
        sock.sendto(final_packet, (ip, port))
        try:
            ack, _ = sock.recvfrom(ACK_SIZE)
            ack_transmission_id, ack_sequence_number = struct.unpack("!HI", ack[:6])
            if (
                ack_transmission_id == transmission_id
                and ack_sequence_number == sequence_number
            ):
                break
        except socket.timeout:
            continue

    sock.close()


def calculate_md5(file_path):
    md5 = hashlib.md5()
    with open(file_path, "rb") as f:
        for chunk in iter(lambda: f.read(4096), b""):
            md5.update(chunk)
    print(md5.hexdigest())
    return md5.digest()


if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: python sender.py <file name> <ip address> <port>")
        sys.exit(1)

    filename = sys.argv[1]
    ip_address = sys.argv[2]
    port = int(sys.argv[3])
    send_file(filename, ip_address, port)

File: send.go
package main

import (
	"crypto/md5"
	"encoding/binary"
	"fmt"
	"io"
	"math/rand"
	"net"
	"os"
	"strconv"
	"sync"
	"time"
)

const (
	PAYLOAD_LENGTH = 1024
	PACKET_SIZE    = PAYLOAD_LENGTH + 6
	WINDOW_SIZE    = 8
	ACK_SIZE       = 6
	TIMEOUT        = 100 * time.Millisecond
)

func sendFile(filePath, ip string, port int) {
	// addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", ip, port))
	// if err != nil {
	// 	panic(err)
	// }
	addr := fmt.Sprintf("%s:%d", ip, port)
	conn, err := net.Dial("udp", addr)
	if err != nil {
		panic(err)
	}
	defer conn.Close()
	transmissionID := uint16(rand.Uint32() % 65535)
	fileInfo, err := os.Stat(filePath)
	if err != nil {
		panic(err)
	}
	fileSize := fileInfo.Size()
	sequenceNumber := uint32(0)
	maxSequenceNumber := uint32(fileSize / PAYLOAD_LENGTH)
	if fileSize%PAYLOAD_LENGTH == 0 {
		maxSequenceNumber += 1
	} else {
		maxSequenceNumber += 2
	}

	zeroPacket := make([]byte, 10+len(filePath))
	binary.BigEndian.PutUint16(zeroPacket[0:2], transmissionID)
	binary.BigEndian.PutUint32(zeroPacket[2:6], sequenceNumber)
	binary.BigEndian.PutUint32(zeroPacket[6:10], maxSequenceNumber)
	copy(zeroPacket[10:], []byte(filePath))

	for {
		conn.Write(zeroPacket)
		ack := make([]byte, ACK_SIZE)
		conn.SetReadDeadline(time.Now().Add(TIMEOUT))
		_, err := conn.Read(ack)
		if err == nil {
			ackTransmissionID := binary.BigEndian.Uint16(ack[0:2])
			ackSequenceNumber := binary.BigEndian.Uint32(ack[2:6])
			if ackTransmissionID == transmissionID && ackSequenceNumber == sequenceNumber {
				fmt.Printf("Ack: %d\n", ackSequenceNumber)
				break
			}
		}
	}
	sequenceNumber++

	var lock sync.Mutex
	base := uint32(1)
	acked := make([]bool, maxSequenceNumber)
	packets := make(map[uint32][]byte)
	var wg sync.WaitGroup

	sendPackets := func() {
		defer wg.Done()
		file, err := os.Open(filePath)
		if err != nil {
			panic(err)
		}
		defer file.Close()

		for base < maxSequenceNumber {
			lock.Lock()
			for sequenceNumber < base+WINDOW_SIZE && sequenceNumber < maxSequenceNumber {
				data := make([]byte, PAYLOAD_LENGTH)
				file.Seek(int64((sequenceNumber-1)*PAYLOAD_LENGTH), io.SeekStart)
				n, err := file.Read(data)
				if err != nil && err != io.EOF {
					panic(err)
				}
				if n == 0 {
					break
				}
				packet := make([]byte, 6+n)
				binary.BigEndian.PutUint16(packet[0:2], transmissionID)
				binary.BigEndian.PutUint32(packet[2:6], sequenceNumber)
				copy(packet[6:], data[:])
				packets[sequenceNumber] = packet
				conn.Write(packet)
				fmt.Printf("Sent: %d\n", sequenceNumber)
				sequenceNumber++
			}
			lock.Unlock()
			time.Sleep(TIMEOUT)
		}
	}

	receiveAcks := func() {
		defer wg.Done()
		ack := make([]byte, ACK_SIZE)
		for base < maxSequenceNumber {
			conn.SetReadDeadline(time.Now().Add(TIMEOUT))
			_, err := conn.Read(ack)
			if err == nil {
				ackTransmissionID := binary.BigEndian.Uint16(ack[0:2])
				ackSequenceNumber := binary.BigEndian.Uint32(ack[2:6])
				fmt.Printf("Ack: %d\n", ackSequenceNumber)
				if ackTransmissionID == transmissionID {
					lock.Lock()
					if !acked[ackSequenceNumber] {
						acked[ackSequenceNumber] = true
						delete(packets, ackSequenceNumber)
						for base < maxSequenceNumber && acked[base] {
							base++
						}
					}
					lock.Unlock()
				}
			} else {
				lock.Lock()
				for seq := base; seq < sequenceNumber; seq++ {
					if !acked[seq] {
						conn.Write(packets[seq])
						fmt.Printf("Resent: %d\n", seq)
					}
				}
				lock.Unlock()
			}
		}
	}

	wg.Add(2)
	go sendPackets()
	go receiveAcks()
	wg.Wait()

	fileMd5 := calculateMD5(filePath)
	finalPacket := make([]byte, 6+len(fileMd5))
	binary.BigEndian.PutUint16(finalPacket[0:2], transmissionID)
	binary.BigEndian.PutUint32(finalPacket[2:6], sequenceNumber)
	copy(finalPacket[6:], fileMd5)
	for {
		conn.Write(finalPacket)
		fmt.Printf("Sent final packet: %d\n", sequenceNumber)
		ack := make([]byte, ACK_SIZE)
		conn.SetReadDeadline(time.Now().Add(TIMEOUT))
		_, err := conn.Read(ack)
		if err == nil {
			ackTransmissionID := binary.BigEndian.Uint16(ack[0:2])
			ackSequenceNumber := binary.BigEndian.Uint32(ack[2:6])
			if ackTransmissionID == transmissionID && ackSequenceNumber == sequenceNumber {
				fmt.Printf("Received final ack: %d\n", ackSequenceNumber)
				break
			}
		}
	}
}

func calculateMD5(filePath string) []byte {
	file, err := os.Open(filePath)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	hash := md5.New()
	if _, err := io.Copy(hash, file); err != nil {
		panic(err)
	}
	return hash.Sum(nil)
}

func main() {
	if len(os.Args) != 4 {
		fmt.Println("Usage: go run send.go <file name> <ip address> <port>")
		os.Exit(1)
	}

	filePath := os.Args[1]
	ip := os.Args[2]
	port, err := strconv.Atoi(os.Args[3])
	if err != nil {
		fmt.Println("Invalid port number")
		os.Exit(1)
	}
	sendFile(filePath, ip, port)
}
